### PR TITLE
add version condition to route even without any routing annotations

### PR DESCRIPTION
**Pull request by dmaicher**
Fixes https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1386 and https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1436

### CODE REVIEW

**Comment by dmaicher** [comment]
Travis CI build fail is unrelated

**Comment by Ener-Getick** [comment]
LGTM :+1:

File: Routing/Loader/Reader/RestActionReader.php [hunk 1]
```diff
@@ -333,39 +333,41 @@ public function read(RestRouteCollection $collection, \ReflectionMethod $method,
...
-                $versionCondition .= '\''.$version.'\'';
-                $first = false;
-            }
-            $versionCondition .= ')';
-            $condition = $condition ? '('.$condition.') and '.$versionCondition : $versionCondition;
+        if ($conditionOne === null) {
+            return $conditionTwo;
+        }
+
+        if ($conditionTwo === null) {
```
Comment (Ener-Getick) [inline Routing/Loader/Reader/RestActionReader.php]: You should use the yoda notation here:
```php
null === $conditionTwo
```
same thing for the condition above

File: Routing/Loader/Reader/RestActionReader.php [hunk 2]
```diff
@@ -333,39 +333,41 @@ public function read(RestRouteCollection $collection, \ReflectionMethod $method,
...
-     * @param RouteAnnotation   $annotation
+     * @return string|null
+     */
+    private function getVersionCondition()
+    {
+        if (empty($this->versions)) {
+            return;
+        }
+
+        return sprintf("request.attributes.get('version') in ['%s']", implode("','", $this->versions));
```
Comment (Ener-Getick) [inline Routing/Loader/Reader/RestActionReader.php]: I'd prefer to add a space after the comma and to use double quotes:
```php
implode('", "', $this->versions)
```

File: Routing/Loader/Reader/RestActionReader.php [hunk 3]
```diff
@@ -333,39 +333,41 @@ public function read(RestRouteCollection $collection, \ReflectionMethod $method,
 
             // add route to collection
             $route = new Route(
-                $path, $defaults, $requirements, $options, $host, [], $methods, $condition
+                $path, $defaults, $requirements, $options, $host, [], $methods, $versionCondition
```
Comment (Ener-Getick) [inline Routing/Loader/Reader/RestActionReader.php]: I'm not sure we should rename this variable, this looks a bit confusing with the one which keeps the same name.

Comment (dmaicher) [inline Routing/Loader/Reader/RestActionReader.php]: Well at this point we are only adding the version condition which was defined in line 295. So I think its actually less confusing like this?

File: Routing/Loader/Reader/RestActionReader.php (hunk 1 above)
Comment (dmaicher) [inline Routing/Loader/Reader/RestActionReader.php]: I personally prefer to not use yoda style but sure I can change it :+1:

File: Routing/Loader/Reader/RestActionReader.php (hunk 2 above)
Comment (dmaicher) [inline Routing/Loader/Reader/RestActionReader.php]: Agree with the space after the comma to improve readability. But why the double quotes? Any reason apart from aesthetics? Just curious :)

File: Routing/Loader/Reader/RestActionReader.php (hunk 3 above)
Comment (dmaicher) [inline Routing/Loader/Reader/RestActionReader.php]: Maybe I can rename `$condition` in line 319 to `$combinedConditions` or similar?

**Comment by dmaicher** [comment]
@Ener-Getick addressed some of your comments. Let me know if something else should be fixed. I kept the single quotes for the expression as they are used like this all over the code ;)

**Comment by Ener-Getick** [comment]
LGTM :+1:

**Comment by Ener-Getick** [comment]
Thank you @dmaicher

**Comment by dmaicher** [comment]
No problem :+1:

File: Tests/Routing/Loader/RestRouteLoaderTest.php [hunk 4]
```diff
@@ -173,7 +173,7 @@ public function testAnnotatedVersionUserFixture()
         $etalonRoutes = $this->loadEtalonRoutesInfo('annotated_version_controller.yml');
 
         $this->assertTrue($collection instanceof RestRouteCollection);
-        $this->assertEquals(2, count($collection->all()));
+        $this->assertEquals(3, count($collection->all()));
```
Comment (stloyd) [inline Tests/Routing/Loader/RestRouteLoaderTest.php]: `assertCount()` should be used instead.

Comment (dmaicher) [inline Tests/Routing/Loader/RestRouteLoaderTest.php]: true could have fixed that :+1:

**Pull request merged**